Nix/Builtins.hs

```diff
@@ -0,0 +1,117 @@
+
+module Nix.Builtins (baseEnv, builtins, evalTopLevelExpr) where
+
+import           Control.Applicative
+import           Control.Monad hiding (mapM, sequence)
+import           Control.Monad.Fix
+import           Data.Fix
+import qualified Data.Map as Map
+import           Data.Text (Text)
+import qualified Data.Text as Text
+import           Data.Traversable (mapM)
+import           Nix.Atoms
+import           Nix.Eval
+import           Nix.Expr (NExpr)
+import           Prelude hiding (mapM, sequence)
+
+-- | Evaluate a nix expression in the default context
+evalTopLevelExpr :: MonadFix m => NExpr -> m (NValue m)
+evalTopLevelExpr val = evalExpr val baseEnv
+
+baseEnv :: MonadFix m => ValueSet m
+baseEnv = Map.fromList $ [ ("builtins", Fix $ NVSet builtins) ] ++ topLevelBuiltins 
+  where
+    topLevelBuiltins = map mapping $ filter isTopLevel builtinsList
+    -- builtins = Map.fromList $ map mapping $ builtinsList
+
+builtins :: MonadFix m => ValueSet m
+builtins = Map.fromList $ map mapping $ builtinsList
+
+data BuiltinType = Normal | TopLevel
+data Builtin m = Builtin {kind :: BuiltinType, mapping :: (Text, NValue m) }
+
+isTopLevel :: Builtin m -> Bool
+isTopLevel b = case kind b of 
+                    Normal -> False
+                    TopLevel -> True
+
+--------------
```

Comment:
It looks like some of this is implemented elsewhere. Can you remove what's been moved, and maybe add a comment explaining why the rest is there?

Reply:
It is implemented right below :-)